src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -0,0 +1,227 @@
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Alipay
+{
+    /// <summary>
+    /// Defines a handler for authentication using Alipay.
+    /// </summary>
+    public class AlipayAuthenticationHandler : OAuthHandler<AlipayAuthenticationOptions>
+    {
+        public AlipayAuthenticationHandler(
+            [NotNull] IOptionsMonitor<AlipayAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
+        {
+            var query = Request.Query;
+            if (query.TryGetValue("auth_code", out var authCode))
+            {
+                // The base `HandleRemoteAuthenticateAsync` requires that `Request.Query` must contain the key called `code`,
+                // which is actually the same as `auth_code` by Alipay's design, but `Request.Query` does not have `Add` operation.
+                // So here is a trick to get the private `Store` dictionary of `QueryCollection`.
+                var queryStore = query.ToDictionary(c => c.Key, c => c.Value, StringComparer.OrdinalIgnoreCase);
+                queryStore["code"] = authCode;
+                Request.QueryString = QueryString.Create(queryStore);
+            }
+
+            return base.HandleRemoteAuthenticateAsync();
+        }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+        {
+            // See https://opendocs.alipay.com/apis/api_9/alipay.system.oauth.token for details.
+            var sortedParams = new SortedDictionary<string, string>()
+            {
+                ["app_id"] = Options.ClientId,
+                ["charset"] = "utf-8",
+                ["code"] = context.Code,
+                ["format"] = "JSON",
+                ["grant_type"] = "authorization_code",
+                ["method"] = "alipay.system.oauth.token",
+                ["sign_type"] = "RSA2",
+                ["timestamp"] = Clock.UtcNow.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture),
```

Comment:
Does Alipay support having `Z` on the end of the timestamp string so you can just use the `"u"` format string instead?

Reply:
As the doc site says, the answer is 'no'.
![image](https://user-images.githubusercontent.com/20787265/109443448-8705d200-7a75-11eb-8bcc-dbedfc4c9cf1.png)